src/base/Macros.hh

```diff
@@ -104,10 +110,63 @@
  * (to provide a more detailed error message in case the point *is* reached).
  */
 #if (!defined(__CUDA_ARCH__) && (defined(__clang__) || defined(__GNUC__))) \
-    || (defined(__CUDA_ARCH__) && CUDART_VERSION >= 11030)
+    || (defined(__CUDA_ARCH__) && CUDART_VERSION >= 11030)                 \
+    || defined(__HIP_DEVICE_COMPILE__)
 #    define CELER_UNREACHABLE __builtin_unreachable()
 #elif defined(_MSC_VER)
 #    define CELER_UNREACHABLE __assume(false)
 #else
 #    define CELER_UNREACHABLE
 #endif
+
+/*!
+ * \def CELER_USE_DEVICE
+ *
+ * True if HIP or CUDA are enabled, false otherwise.
+ */
+#if CELERITAS_USE_CUDA || CELERITAS_USE_HIP
+#    define CELER_USE_DEVICE 1
+#else
+#    define CELER_USE_DEVICE 0
+#endif
+
+/*!
+ * \def CELER_DEVICE_SOURCE
+ *
+ * Defined and true if building a HIP or CUDA source file. This is a generic
+ * replacement for \c __CUDACC__ .
+ */
+#if defined(__CUDACC__) || defined(__HIP__)
+#    define CELER_DEVICE_SOURCE 1
+#endif
+
+/*!
+ * \def CELER_DEVICE_COMPILE
+ *
+ * Defined and true if building device code in HIP or CUDA. This is a generic
+ * replacement for \c #ifdef \c __CUDA_ARCH__ .
+ */
+#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+#    define CELER_DEVICE_COMPILE 1
+#endif
+
+/*!
+ * \def CELER_DEVICE_PREFIX
+ *
+ * Add a prefix "hip", "cuda", or "mock" to a code token.
+ */
+/*!
+ * \def CELER_DEVICE_SHORT_PREFIX
+ *
+ * Add a prefix "hip", "cu", or "mock" to a code token.
+ */
+#if CELERITAS_USE_CUDA
+#    define CELER_DEVICE_PREFIX(TOK) cuda##TOK
+#    define CELER_DEVICE_SHORT_PREFIX(TOK) cu##TOK
```

Comment:
I confess to not really loving the `SHORT_PREFIX`.  One of this issues with this macro approach is that it potentially obfuscates runtime (host-side) versus device-side interfaces. However, given how tightly HIP is following CUDA, this probably isn't ever going to be a problem. 

The only issue down stream is that things like `cudaMemcpyHostToDevice` which are currently enumerations, could become legitimate types in a future version while HIP remains enumerations.  This probably still won't cause any problems with the model you are using though.

What's more likely (if anything) to occur is that some device-side interfaces require some extra changes that fall outside the MACRO model to handle them.  But I would worry about that on a case-by-case basis, and wouldn't let that stop me from merging at this point.

Reply:
It's definitely a yucky expedient... since it's only about changing the names of symbols I think it won't cause any problems. The only situation it could break I think is if someone at NVIDIA/AMD decides that `cudaDeviceSynchronize` should be a macro...